src/runner_manager.py

```diff
@@ -0,0 +1,358 @@
+from runner import Runner
+from runner_type import GitHubOrg, GitHubRepo, ProxySetting, VirtualMachineResources
+from utilities import execute_command, retry
+
+logger = logging.getLogger(__name__)
+
+
+@dataclass
+class RunnerInfo:
+    """Information from GitHub of a runner.
+
+    Used as a returned type to method querying runner information.
+
+    TODO:
+        See if more information should be shared by the charm.
+    """
+
+    name: str
+    online: bool
+    offline: bool
+
+
+class RunnerManager:
+
+    # TODO: Verify if this violates bandit B108.
+    runner_bin_path = Path("/var/cache/github-runner-operator/runner.tgz")
+
+    def __init__(
+        self, path: str, token: str, app_name: str, reconcile_interval: int, image: str = "focal"
+    ) -> None:
+        """Construct RunnerManager object for creating and managing runners.
+
+        Args:
+            path: GitHub repository path in the format '<owner>/<repo>', or the GitHub organization
+                name.
+            token: GitHub personal access token to register runner to the repository or
+                organization.
+            app_name: An name for the set of runners.
+            reconcile_interval: Number of minutes between each reconciliation of runners.
+            image: Image to use for the runner LXD instances.
+
+        TODO:
+            Move `install_deps` class method elsewhere.
+
+            Change the `path` in `__init__` method to type `GitHubPath`.
+        """
+        http_proxy = os.environ.get("JUJU_CHARM_HTTP_PROXY", None)
+        https_proxy = os.environ.get("JUJU_CHARM_HTTPS_PROXY", None)
+        no_proxy = os.environ.get("JUJU_CHARM_NO_PROXY", None)
+
+        self.proxies: ProxySetting = {}
+        if http_proxy:
+            self.proxies["http"] = http_proxy
+        if https_proxy:
+            self.proxies["https"] = https_proxy
+        if no_proxy:
+            self.proxies["no_proxy"] = no_proxy
+
+        self.session = requests.Session()
+        # TODO: Review the retry strategy
+        adapter = requests.adapters.HTTPAdapter(
+            max_retries=urllib3.Retry(
+                total=10, backoff_factor=0.3, status_forcelist=[500, 502, 503, 504]
+            )
+        )
+        self.session.mount("http://", adapter)
+        self.session.mount("https://", adapter)
+        if self.proxies:
+            # setup proxy for requests
+            self.session.proxies.update(self.proxies)
+            # add proxy to fastcore which ghapi uses
+            proxy = urllib.request.ProxyHandler(self.proxies)
+            opener = urllib.request.build_opener(proxy)
+            fastcore.net._opener = opener
+
+        self._github = GhApi(token=token)
+        self._jinja = jinja2.Environment(
+            loader=jinja2.FileSystemLoader("templates"), autoescape=True
+        )
+        self._lxd = pylxd.Client()
+
+        if "/" in path:
+            owner, repo = path.split("/")
```

Comment:
What is the user has put more than 1 `/` in the path? Should we check that input first?